src/pc/1.17/ChunkColumn.js

```diff
@@ -0,0 +1,275 @@
+const SmartBuffer = require('smart-buffer').SmartBuffer
+const ChunkSection = require('./ChunkSection')
+const constants = require('../common/constants')
+const varInt = require('../common/varInt')
+const BitArray = require('../common/BitArrayNoSpan')
+
+// wrap with func to provide version specific Block
+module.exports = (Block, mcData) => {
+  return class ChunkColumn {
+    constructor (options) {
+      this.minWorldHeight = options.minWorldHeight ?? 0
+      this.maxWorldHeight = options.maxWorldHeight ?? constants.CHUNK_HEIGHT
+
+      this.sectionMask = 0
+      this.sections = Array(this.countVerticalSections()).fill(null)
+      this.biomes = Array(4 * 4 * this.countVerticalBiomeGridSize()).fill(0)
+    }
+
+    initialize (func) {
+      const p = { x: 0, y: 0, z: 0 }
+      for (p.y = this.minWorldHeight; p.y < this.maxWorldHeight; p.y++) {
+        for (p.z = 0; p.z < constants.SECTION_WIDTH; p.z++) {
+          for (p.x = 0; p.x < constants.SECTION_WIDTH; p.x++) {
+            const block = func(p.x, p.y, p.z)
+            if (block === null) {
+              continue
+            }
+            this.setBlock(p, block)
+          }
+        }
+      }
+    }
+
+    // Json serialization/deserialization methods
+    toJson () {
+      return JSON.stringify({
+        minWorldHeight: this.minWorldHeight,
+        maxWorldHeight: this.maxWorldHeight,
+        biomes: this.biomes,
+        sectionMask: this.sectionMask,
+        sections: this.sections.map(section => section === null ? null : section.toJson())
+      })
+    }
+
+    static fromJson (j) {
+      const parsed = JSON.parse(j)
+      const chunk = new ChunkColumn()
+      chunk.biomes = parsed.biomes
+      chunk.sectionMask = parsed.sectionMask
+      chunk.sections = parsed.sections.map(s => s === null ? null : ChunkSection.fromJson(s))
+
+      if (parsed.minWorldHeight) {
+        chunk.minWorldHeight = parsed.minWorldHeight
+      }
+      if (parsed.maxWorldHeight) {
+        chunk.maxWorldHeight = parsed.maxWorldHeight
+      }
+      return chunk
+    }
+
+    // Utility methods used for retrieving section indices and checking block positions
+    getMask () {
+      return this.sectionMask
+    }
+
+    countVerticalBiomeGridSize () {
+      const worldTotalHeight = this.maxWorldHeight - this.minWorldHeight
+      return Math.floor((worldTotalHeight + 3) / 4)
+    }
+
+    countVerticalSections () {
```

Comment:
This function calls a lot of tiny functions that could be inlined into:
```js
countVerticalSections () {
  return (this.maxWorldHeight - this.minWorldHeight) >> 4
}
```
Function calls are not free.

Reply:
That particular function is called exactly once per chunk creation and twice for chunk load, performance claim just makes no sense here.

Reply:
But alright, I inlined it since that way it actually looks simpler